File: scripts/monitoring/notifier.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

require 'json'
require 'fileutils'
require 'time'
require 'optparse'
require 'net/http'
require 'uri'

module Gmux
  module Scripts
    class Notifier
      STORE_PATH = File.join(Dir.home, '.gmux', 'sessions.json')

      def initialize(options = {})
        @interval = options[:interval] || 5
        @verbose = options[:verbose] || false
        @webhook = options[:webhook]
        @email = options[:email]
        @slack = options[:slack]
        @discord = options[:discord]
        @telegram = options[:telegram]
        @sound = options[:sound] || false
        @watch_session = options[:session]
        @notify_on = options[:notify_on] || ['complete', 'error']
      end

      def run
        puts "=== GMUX Session Notifier ==="
        puts "Interval: #{@interval}s"
        puts "Notify on: #{@notify_on.join(', ')}"
        puts "Methods: #{notification_methods.join(', ')}"
        puts

        watch_sessions
      end

      private

      def notification_methods
        methods = []
        methods << 'webhook' if @webhook
        methods << 'email' if @email
        methods << 'slack' if @slack
        methods << 'discord' if @discord
        methods << 'telegram' if @telegram
        methods << 'sound' if @sound
        methods << 'stdout' if methods.empty?
        methods
      end

      def load_sessions
        return {} unless File.exist?(STORE_PATH)
        JSON.parse(File.read(STORE_PATH))
      rescue JSON::ParserError
        {}
      end

      def watch_sessions
        puts "Watching for session completion... (Ctrl+C to stop)"
        puts

        running = true
        trap("INT") { running = false }

        @last_status = {}

        while running
          sessions = load_sessions
          sessions_to_watch = @watch_session ?
            sessions.select { |name, _| name == @watch_session } :
            sessions

          sessions_to_watch.each do |name, session|
            current_status = session['status']
            last_status = @last_status[name]

            # Check if status changed to a notify-worthy status
            if last_status && last_status != current_status && @notify_on.include?(current_status)
              notify(name, session, last_status)
            end

            @last_status[name] = current_status
          end

          sleep @interval
        end

        puts "\nStopped watching."
      end

      def notify(session_name, session, previous_status)
        message = build_message(session_name, session, previous_status)

        puts if @verbose
        puts "  Notification: #{session_name} is now #{session['status']}" if @verbose

        # Send notifications via all configured methods
        send_webhook(message) if @webhook
        send_email(message) if @email
        send_slack(message) if @slack
        send_discord(message) if @discord
        send_telegram(message) if @telegram
        play_sound if @sound

        # Always print to stdout if no other method configured
        unless @webhook || @email || @slack || @discord || @telegram
          display_notification(message)
        end
      end

      def build_message(session_name, session, previous_status)
        {
          session: session_name,
          status: session['status'],
          previous_status: previous_status,
          agent: session['agentCommand'],
          branch: session['branchName'],
          worktree: session['worktreePath'],
          started: session['startedAt'],
          timestamp: Time.now.iso8601
        }
      end

      def display_notification(message)
        puts
        puts "=" * 50
        puts "SESSION NOTIFICATION"
        puts "=" * 50
        puts "Session: #{message[:session]}"
        puts "Status: #{message[:status].upcase}"
        puts "Previous: #{message[:previous_status]}"
        puts "Agent: #{message[:agent]}"
        puts "Branch: #{message[:branch]}"
        puts "Time: #{message[:timestamp]}"
        puts "=" * 50
      end

      def send_webhook(message)
        uri = URI.parse(@webhook)
        http = Net::HTTP.new(uri.host, uri.port)
        http.use_ssl = (uri.scheme == 'https')

        request = Net::HTTP::Post.new(uri.path, { 'Content-Type' => 'application/json' })
        request.body = {
          text: "GMUX Session #{message[:status].upcase}",
          session: message[:session],
          status: message[:status],
          agent: message[:agent],
          timestamp: message[:timestamp]
        }.to_json

        begin
          response = http.request(request)
          puts "  Webhook sent: #{response.code}" if @verbose
        rescue StandardError => e
          puts "  Webhook error: #{e.message}"
        end
      end

      def send_email(message)
        # Basic email using mail command (if available)
        subject = "GMUX: #{message[:session]} is #{message[:status]}"
        body = <<~BODY
          Session: #{message[:session]}
          Status: #{message[:status]}
          Agent: #{message[:agent]}
          Branch: #{message[:branch]}
          Time: #{message[:timestamp]}
        BODY

        if @email.include?('@')
          # Send to email address
          system("echo '#{body}' | mail -s '#{subject}' #{@email}")
          puts "  Email sent to: #{@email}" if @verbose
        else
          # Use as mail command
          system("echo '#{body}' | #{@email} -s '#{subject}'")
          puts "  Email sent via: #{@email}" if @verbose
        end
      end

      def send_slack(message)
        slack_webhook = @slack
        uri = URI.parse(slack_webhook)

        payload = {
          text: "GMUX Session #{message[:status].upcase}",
          attachments: [
            {
              color: message[:status] == 'complete' ? 'good' : 'danger',
              fields: [
                { title: 'Session', value: message[:session], short: true },
                { title: 'Status', value: message[:status], short: true },
                { title: 'Agent', value: message[:agent], short: true },
                { title: 'Branch', value: message[:branch], short: true }
              ],
              timestamp: Time.now.to_i
            }
          ]
        }

        http = Net::HTTP.new(uri.host, uri.port)
        http.use_ssl = (uri.scheme == 'https')
        request = Net::HTTP::Post.new(uri.path, { 'Content-Type' => 'application/json' })
        request.body = payload.to_json

        begin
          response = http.request(request)
          puts "  Slack sent: #{response.code}" if @verbose
        rescue StandardError => e
          puts "  Slack error: #{e.message}"
        end
      end

      def send_discord(message)
        discord_webhook = @discord
        uri = URI.parse(discord_webhook)

        payload = {
          content: "GMUX Session #{message[:status].upcase}",
          embeds: [
            {
              title: "Session #{message[:status].capitalize}",
              description: "Session `#{message[:session]}` has #{message[:status]}.",
              color: message[:status] == 'complete' ? 0x00ff00 : 0xff0000,
              fields: [
                { name: 'Agent', value: message[:agent], inline: true },
                { name: 'Branch', value: message[:branch], inline: true },
                { name: 'Time', value: message[:timestamp], inline: false }
              ]
            }
          ]
        }

        http = Net::HTTP.new(uri.host, uri.port)
        http.use_ssl = (uri.scheme == 'https')
        request = Net::HTTP::Post.new(uri.path, { 'Content-Type' => 'application/json' })
        request.body = payload.to_json

        begin
          response = http.request(request)
          puts "  Discord sent: #{response.code}" if @verbose
        rescue StandardError => e
          puts "  Discord error: #{e.message}"
        end
      end

      def send_telegram(message)
        telegram_bot, chat_id = @telegram.split(':')
        return unless telegram_bot && chat_id

        uri = URI.parse("https://api.telegram.org/bot#{telegram_bot}/sendMessage")

        text = <<~TEXT
          *GMUX Session #{message[:status].upcase}*
          
          Session: #{message[:session]}
          Status: #{message[:status]}
          Agent: #{message[:agent]}
          Branch: #{message[:branch]}
          Time: #{message[:timestamp]}
        TEXT

        payload = {
          chat_id: chat_id,
          text: text,
          parse_mode: 'Markdown'
        }

        http = Net::HTTP.new(uri.host, uri.port)
        http.use_ssl = true
        request = Net::HTTP::Post.new(uri.path, { 'Content-Type' => 'application/json' })
        request.body = payload.to_json

        begin
          response = http.request(request)
          puts "  Telegram sent: #{response.code}" if @verbose
        rescue StandardError => e
          puts "  Telegram error: #{e.message}"
        end
      end

      def play_sound
        # Play a notification sound (macOS)
        if system('which afplay > /dev/null 2>&1')
          system('afplay /System/Library/Sounds/Glass.aiff')
        elsif system('which aplay > /dev/null 2>&1')
          system('aplay /usr/share/sounds/freedesktop/stereo/complete.oga')
        end
      end
    end
  end
end

if __FILE__ == $PROGRAM_NAME
  options = {}
  OptionParser.new do |opts|
    opts.banner = "Usage: #{$PROGRAM_NAME} [options]"

    opts.on('-i', '--interval SECONDS', Integer, "Check interval in seconds (default: 5)") do |i|
      options[:interval] = i
    end

    opts.on('-s', '--session NAME', "Watch specific session") do |s|
      options[:session] = s
    end

    opts.on('-w', '--webhook URL', "Send webhook notification") do |w|
      options[:webhook] = w
    end

    opts.on('--slack URL', "Send Slack notification") do |s|
      options[:slack] = s
    end

    opts.on('--discord URL', "Send Discord notification") do |d|
      options[:discord] = d
    end

    opts.on('--telegram BOT:CHAT_ID', "Send Telegram notification") do |t|
      options[:telegram] = t
    end

    opts.on('-e', '--email ADDRESS', "Send email notification") do |e|
      options[:email] = e
    end

    opts.on('--sound', "Play sound on notification") do
      options[:sound] = true
    end

    opts.on('-n', '--notify-on STATUS', Array, "Notify on status (default: complete,error)") do |n|
      options[:notify_on] = n
    end

    opts.on('-v', '--verbose', "Show detailed output") do
      options[:verbose] = true
    end

    opts.on('-h', '--help', "Show this help") do
      puts opts
      exit
    end
  end.parse!

  Gmux::Scripts::Notifier.new(options).run
end

File: scripts/monitoring/notifier.py
#!/usr/bin/env python3
"""GMUX Session Notifier - Send notifications on session events."""

import json
import logging
import os
import signal
import subprocess
import sys
import time
from argparse import ArgumentParser
from datetime import datetime
from pathlib import Path
from typing import Any, Dict, List, Optional
from urllib.parse import urlparse
from urllib.request import Request, urlopen

logging.basicConfig(level=logging.INFO, format="%(message)s")
logger = logging.getLogger(__name__)

STORE_PATH = Path.home() / ".gmux" / "sessions.json"


class Notifier:
    def __init__(self, options: Dict[str, Any]):
        self.interval = options.get("interval", 5)
        self.verbose = options.get("verbose", False)
        self.webhook = options.get("webhook")
        self.email = options.get("email")
        self.slack = options.get("slack")
        self.discord = options.get("discord")
        self.telegram = options.get("telegram")
        self.sound = options.get("sound", False)
        self.watch_session = options.get("session")
        self.notify_on = options.get("notify_on", ["complete", "error"])
        self.running = True
        self.last_status = {}

        signal.signal(signal.SIGINT, self._handle_interrupt)

    def _handle_interrupt(self, signum, frame):
        self.running = False

    def run(self):
        print("=== GMUX Session Notifier ===")
        print(f"Interval: {self.interval}s")
        print(f"Notify on: {', '.join(self.notify_on)}")
        print(f"Methods: {', '.join(self._notification_methods())}")
        print()

        self.watch_sessions()

    def _notification_methods(self) -> List[str]:
        methods = []
        if self.webhook:
            methods.append("webhook")
        if self.email:
            methods.append("email")
        if self.slack:
            methods.append("slack")
        if self.discord:
            methods.append("discord")
        if self.telegram:
            methods.append("telegram")
        if self.sound:
            methods.append("sound")
        if not methods:
            methods.append("stdout")
        return methods

    def load_sessions(self) -> Dict[str, Any]:
        if not STORE_PATH.exists():
            return {}
        try:
            return json.loads(STORE_PATH.read_text())
        except (json.JSONDecodeError, OSError):
            return {}

    def watch_sessions(self):
        print("Watching for session completion... (Ctrl+C to stop)")
        print()

        while self.running:
            sessions = self.load_sessions()
            sessions_to_watch = (
                {k: v for k, v in sessions.items() if k == self.watch_session}
                if self.watch_session
                else sessions
            )

            for name, session in sessions_to_watch.items():
                current_status = session.get("status")
                last_status = self.last_status.get(name)

                if (
                    last_status
                    and last_status != current_status
                    and current_status in self.notify_on
                ):
                    self._notify(name, session, last_status)

                self.last_status[name] = current_status

            time.sleep(self.interval)

        print("\nStopped watching.")

    def _notify(self, session_name: str, session: Dict[str, Any], previous_status: str):
        message = self._build_message(session_name, session, previous_status)

        if self.verbose:
            print()
            print(f"  Notification: {session_name} is now {session.get('status')}")

        # Send via all configured methods
        if self.webhook:
            self._send_webhook(message)
        if self.email:
            self._send_email(message)
        if self.slack:
            self._send_slack(message)
        if self.discord:
            self._send_discord(message)
        if self.telegram:
            self._send_telegram(message)
        if self.sound:
            self._play_sound()

        # Print to stdout if no other method configured
        if not any([self.webhook, self.email, self.slack, self.discord, self.telegram]):
            self._display_notification(message)

    def _build_message(
        self, session_name: str, session: Dict[str, Any], previous_status: str
    ) -> Dict[str, Any]:
        return {
            "session": session_name,
            "status": session.get("status"),
            "previous_status": previous_status,
            "agent": session.get("agentCommand"),
            "branch": session.get("branchName"),
            "worktree": session.get("worktreePath"),
            "started": session.get("startedAt"),
            "timestamp": datetime.now().isoformat(),
        }

    def _display_notification(self, message: Dict[str, Any]):
        print()
        print("=" * 50)
        print("SESSION NOTIFICATION")
        print("=" * 50)
        print(f"Session: {message['session']}")
        print(f"Status: {message['status'].upper()}")
        print(f"Previous: {message['previous_status']}")
        print(f"Agent: {message['agent']}")
        print(f"Branch: {message['branch']}")
        print(f"Time: {message['timestamp']}")
        print("=" * 50)

    def _send_webhook(self, message: Dict[str, Any]):
        try:
            payload = {
                "text": f"GMUX Session {message['status'].upper()}",
                "session": message["session"],
                "status": message["status"],
                "agent": message["agent"],
                "timestamp": message["timestamp"],
            }

            req = Request(
                self.webhook,
                data=json.dumps(payload).encode("utf-8"),
                headers={"Content-Type": "application/json"},
                method="POST",
            )

            with urlopen(req, timeout=5) as response:
                if self.verbose:
                    print(f"  Webhook sent: {response.status}")
        except Exception as e:
            logger.error(f"  Webhook error: {e}")

    def _send_email(self, message: Dict[str, Any]):
        subject = f"GMUX: {message['session']} is {message['status']}"
        body = f"""Session: {message['session']}
Status: {message['status']}
Agent: {message['agent']}
Branch: {message['branch']}
Time: {message['timestamp']}
"""

        try:
            if "@" in self.email:
                # Send to email address
                cmd = f"echo '{body}' | mail -s '{subject}' {self.email}"
            else:
                # Use as mail command
                cmd = f"echo '{body}' | {self.email} -s '{subject}'"

            subprocess.run(cmd, shell=True, check=False)
            if self.verbose:
                print(f"  Email sent to: {self.email}")
        except Exception as e:
            logger.error(f"  Email error: {e}")

    def _send_slack(self, message: Dict[str, Any]):
        try:
            payload = {
                "text": f"GMUX Session {message['status'].upper()}",
                "attachments": [
                    {
                        "color": "good" if message["status"] == "complete" else "danger",
                        "fields": [
                            {"title": "Session", "value": message["session"], "short": True},
                            {"title": "Status", "value": message["status"], "short": True},
                            {"title": "Agent", "value": message["agent"], "short": True},
                            {"title": "Branch", "value": message["branch"], "short": True},
                        ],
                        "ts": int(datetime.now().timestamp()),
                    }
                ],
            }

            req = Request(
                self.slack,
                data=json.dumps(payload).encode("utf-8"),
                headers={"Content-Type": "application/json"},
                method="POST",
            )

            with urlopen(req, timeout=5) as response:
                if self.verbose:
                    print(f"  Slack sent: {response.status}")
        except Exception as e:
            logger.error(f"  Slack error: {e}")

    def _send_discord(self, message: Dict[str, Any]):
        try:
            payload = {
                "content": f"GMUX Session {message['status'].upper()}",
                "embeds": [
                    {
                        "title": f"Session {message['status'].capitalize()}",
                        "description": f"Session `{message['session']}` has {message['status']}.",
                        "color": 0x00FF00 if message["status"] == "complete" else 0xFF0000,
                        "fields": [
                            {"name": "Agent", "value": message["agent"], "inline": True},
                            {"name": "Branch", "value": message["branch"], "inline": True},
                            {"name": "Time", "value": message["timestamp"], "inline": False},
                        ],
                    }
                ],
            }

            req = Request(
                self.discord,
                data=json.dumps(payload).encode("utf-8"),
                headers={"Content-Type": "application/json"},
                method="POST",
            )

            with urlopen(req, timeout=5) as response:
                if self.verbose:
                    print(f"  Discord sent: {response.status}")
        except Exception as e:
            logger.error(f"  Discord error: {e}")

    def _send_telegram(self, message: Dict[str, Any]):
        try:
            parts = self.telegram.split(":")
            if len(parts) != 2:
                logger.error("Invalid Telegram format. Use: BOT_TOKEN:CHAT_ID")
                return

            bot_token, chat_id = parts
            url = f"https://api.telegram.org/bot{bot_token}/sendMessage"

            text = f"""*GMUX Session {message['status'].upper()}*

Session: {message['session']}
Status: {message['status']}
Agent: {message['agent']}
Branch: {message['branch']}
Time: {message['timestamp']}
"""

            payload = {
                "chat_id": chat_id,
                "text": text,
                "parse_mode": "Markdown",
            }

            req = Request(
                url,
                data=json.dumps(payload).encode("utf-8"),
                headers={"Content-Type": "application/json"},
                method="POST",
            )

            with urlopen(req, timeout=5) as response:
                if self.verbose:
                    print(f"  Telegram sent: {response.status}")
        except Exception as e:
            logger.error(f"  Telegram error: {e}")

    def _play_sound(self):
        try:
            # Try macOS sound
            if subprocess.run(["which", "afplay"], capture_output=True).returncode == 0:
                subprocess.run(
                    ["afplay", "/System/Library/Sounds/Glass.aiff"],
                    check=False,
                )
            # Try Linux sound
            elif subprocess.run(["which", "aplay"], capture_output=True).returncode == 0:
                subprocess.run(
                    ["aplay", "/usr/share/sounds/freedesktop/stereo/complete.oga"],
                    check=False,
                )
        except Exception as e:
            logger.debug(f"Sound error: {e}")


def main():
    parser = ArgumentParser(description="Send GMUX session notifications")
    parser.add_argument(
        "-i",
        "--interval",
        type=int,
        default=5,
        help="Check interval in seconds (default: 5)",
    )
    parser.add_argument("-s", "--session", help="Watch specific session")
    parser.add_argument("-w", "--webhook", help="Send webhook notification")
    parser.add_argument("--slack", help="Send Slack notification")
    parser.add_argument("--discord", help="Send Discord notification")
    parser.add_argument("--telegram", help="Send Telegram notification (BOT:CHAT_ID)")
    parser.add_argument("-e", "--email", help="Send email notification")
    parser.add_argument("--sound", action="store_true", help="Play sound on notification")
    parser.add_argument(
        "-n",
        "--notify-on",
        nargs="+",
        default=["complete", "error"],
        help="Notify on status (default: complete error)",
    )
    parser.add_argument("-v", "--verbose", action="store_true", help="Show detailed output")

    args = parser.parse_args()

    notifier = Notifier(vars(args))
    notifier.run()


if __name__ == "__main__":
    main()
